### PR TITLE
[SPARK-49152][SQL][FOLLOWUP] table location string should be Hadoop Path string

### PR DESCRIPTION
**Pull request by cloud-fan**
### What changes were proposed in this pull request?

This is a followup of https://github.com/apache/spark/pull/47660 to restore the behavior change. The table location string should be Hadoop Path string instead of URL string which escapes all special chars.

### Why are the changes needed?

restore the unintentional behavior change.

### Does this PR introduce _any_ user-facing change?

No, it's not released yet

### How was this patch tested?

new test

### Was this patch authored or co-authored using generative AI tooling?

no

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveSessionCatalog.scala [hunk 1]
```diff
@@ -373,8 +373,15 @@ class ResolveSessionCatalog(val catalogManager: CatalogManager)
         serdeProperties,
         partitionSpec)
 
-    case SetTableLocation(ResolvedV1TableIdentifier(ident), partitionSpec, location) =>
-      AlterTableSetLocationCommand(ident, partitionSpec, location)
+    case SetTableLocation(ResolvedV1TableIdentifier(ident), None, location) =>
+      AlterTableSetLocationCommand(ident, None, location)
+
+    // V2 catalog doesn't support setting partition location yet, we must use v1 command here.
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveSessionCatalog.scala:379]: not related to table location but is also a followup of https://github.com/apache/spark/pull/47660 to fix a missing case.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DataSourceV2Strategy.scala [hunk 2]
```diff
@@ -89,8 +91,19 @@ class DataSourceV2Strategy(session: SparkSession) extends Strategy with Predicat
   }
 
   private def qualifyLocInTableSpec(tableSpec: TableSpec): TableSpec = {
-    tableSpec.withNewLocation(tableSpec.location.map(loc => CatalogUtils.makeQualifiedPath(
-      CatalogUtils.stringToURI(loc), hadoopConf).toString))
+    val newLoc = tableSpec.location.map { loc =>
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DataSourceV2Strategy.scala:94]: The code here follows `SessionCatalog#makeQualifiedTablePath`

**Review by dongjoon-hyun — APPROVED** [review]
+1, this looks correct to me.

File: sql/core/src/test/resources/sql-tests/results/show-create-table.sql.out [hunk 3]
```diff
@@ -78,7 +78,7 @@ CREATE TABLE spark_catalog.default.tbl (
   b STRING,
   c INT)
 USING parquet
-LOCATION 'file:///path/to/table'
+LOCATION 'file:/path/to/table'
```
Comment (cloud-fan) [inline sql/core/src/test/resources/sql-tests/results/show-create-table.sql.out:81]: I've investigated this locally. The new result is actually consistent with the production behavior (with Hive Metastore). What happens is:
- CREATE TABLE command qualifies the table location, and keeps it as `URI` in `CatalogTable#storage#locationUri`
- When saving the table into HMS, we have to turn `URI` into string, by using `new Path(uri).toString`. The Hadoop Path string will omit the `authority` component if it's not present. So `file:///path` becomes `file:/path`.
- When reading the table back from HMS, we turn string back to `URI`, but the `authority` component won't be filled.

However, with `InMemoryCatalog`, we keep `CatalogTable` directly and there is no `URI` <-> string round trip. So the `authority` component is still there.

This actually doesn't matter, as empty `authority` is the same as no `authority`, in `URI` string.

Comment (dongjoon-hyun) [reply sql/core/src/test/resources/sql-tests/results/show-create-table.sql.out:81]: Thank you for the details.

**Comment by dongjoon-hyun** [comment]
Merged to master.

**Comment by dongjoon-hyun** [comment]
Could you make a backporting PR to branch-3.5 seperately in order to make it sure all CIs pass, @cloud-fan ?

**Pull request closed without merging**